keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,1114 @@
+// Each entry includes:
+//
+//      IP: An IP address allocated to the pod.
+//      PodNetworkName: Name of the PodNetwork the IP belongs to.
+//      InterfaceName: Name of the network interface inside the Pod.
+type PodIP struct {
+        // ip is an IP address (IPv4 or IPv6) assigned to the pod
+        IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
+
+        // PodNetworkName is name of the PodNetwork the IP belongs to
+        //
+        // +optional
+        PodNetworkName string `json:"podNetwork"`
+
+        // InterfaceName is name of the network interface used for this attachment
+        //
+        // +optional
+        InterfaceName string `json:"interfaceName",omitempty`
+}
+```
+Example:
+```yaml
+kind: Pod
+metadata:
+  name: pod1
+  namespace: default
+spec:
+[...]
+  networks:
+  - podNetwork: default
+    interfaceName: eth0
+    isDefaultGW4: true
+  - podNetwork: dataplane1
+    interfaceName: net1
+  - podNetworkAttachmentName: my-interface-fast
+    interfaceName: net2
+status:
+[...]
+  podIP: 192.168.5.54
+  podIPs:
+  - ip: 192.168.5.54
+    podNetwork: default
+    interfaceName: eth0
+  - ip: 10.0.0.20
+    podNetwork: dataplane1
+    interfaceName: net1
+  - ip: 2011::233
+    podNetwork: my-interface-fast
+    interfaceName: net2
+```
+The above status is expected to be populated by kubelet, but this can only happen
+after CRI provides support for the new Pod API. Because of that, initially
+kubelet will behave as it does today, without updating the additional fields.
+Until CRI catches up, the PodNetwork providers will be able to update that field
+on their own.
+
+#### DRA integration (alternative)
+We have discussed, as an alternative model, usage of the DRA API, and concluded
+that using it will be less clear for the user, compared to the explicit
+PodNetwork model, proposed above.
+
+### API server changes
+These are the changes for the API server covered by this design:
+* Provide validation webhook for PodNetwork or PodNetworkAttachment objects
+* Handle “default” PodNetwork deletion
+* Extend static Pod spec validation
+
+### Scheduler changes
+These are the changes we will do in Pod scheduler:
+* Provide active Pod spec validation
+
+When one of the multi-network validation fails, scheduler will follow the current
+“failure” path for Pod:
+* set PodScheduled condition (of the Pod) to False with appropriate error message
+* send Pod Event with same error message
+
+### Endpointslice controller changes
+Considering changes to Pod.PodStatus.PodIPs list, we must ensure that the
+controller is using the correct IPs when creating the endpoints. We will ensure
+that only IPs for "defautl" PodNetwork will be used.
+
+### Kubelet changes
+We will introduce an additional check for kubelet readiness for networking.
+Today kubelet does this verification via CRI that checks CNI config presence.
+We will add a “default” PodNetwork presence check to this flow. Until such a
```

Comment:
This requires more detail, to do kubelet do a CRI runtime check using a RuntimeCondition RCP call, and this is replied by the container runtime, both crio and containerd use the presence of the cni config file to indicate the readiness of the network.

With this model, this has to be defined more clearly
- what happens if a kubelet has 2 networks and one is ready and the other no?
- how do the kubelet behave with networks that are created after the kubelet starts?
- how kubelet standalone works, if it does not have way to check the pod network objects?

Reply:
This section is being removed as per above feedback.